### PR TITLE
gh-231: add `ruff` linting to `pre-commit`

### LINKED ISSUE #231: Add `ruff` linting
### Is Your Feature Request Related to a Problem? Please Describe

Following #199, adding in a separate PR to separate linting and formatting

### Describe the Solution You'd Like

_No response_

### Describe Alternatives You've Considered

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by paddyroddy**
Add `ruff` linting and turn on as many rules as possible. However, I will turn off many and not spend too much time fixing them. In a perfect world, we will try and fix as many of these as possible - as they will generally help with maintenance in the long term.

Closes: #231
Refs: #170

### CODE REVIEW

File: glass/lensing.py [hunk 1]
```diff
@@ -67,6 +72,8 @@ def from_convergence(
         Maximum angular mode number to use in the transform.
     potential, deflection, shear : bool, optional
         Which lensing maps to return.
+    discretized : bool
+        Whether to apply a pixel window function to the shear.
```
Comment (paddyroddy) [inline glass/lensing.py:76]: @ntessore this was missing, I've had a crack. Please feel free to edit.

File: pyproject.toml [hunk 2]
```diff
@@ -94,6 +94,58 @@ force-exclude = true
...
+    "future",
+    "standard-library",
+    "third-party",
+    "cosmo",
+    "first-party",
+    "local-folder",
+], sections = {"cosmo" = [
+    "camb",
+    "cosmology",
+]}}
```
Comment (paddyroddy) [inline pyproject.toml:123]: This has the effect of grouping `camb` and `cosmology` after third-party stuff, but before `glass`. Do we want anything else here? @ntessore

File: .git-blame-ignore-revs [hunk 3]
```diff
@@ -2,3 +2,4 @@
 b42067f6776dcd763827000d585a88e071b78af3
 # applied prettier - https://github.com/glass-dev/glass/pull/227
 f9bac62f497a7288aa71fd4dbd948945c37f854e
+# applied ruff-linting - https://github.com/glass-dev/glass/pull/232
```
Comment (paddyroddy) [inline .git-blame-ignore-revs:5]: @Saransh-cpp when this is merged can you add the commit hash underneath?

File: glass/lensing.py (hunk 1 above)
Comment (ntessore) [reply glass/lensing.py:76]: Close!
```suggestion
    discretized : bool
        Correct the pixel window function in output maps.
```

File: .git-blame-ignore-revs (hunk 3 above)
Comment (Saransh-cpp) [reply .git-blame-ignore-revs:5]: 🫡

File: pyproject.toml (hunk 2 above)
Comment (ntessore) [reply pyproject.toml:123]: Looks good to me! Very cool

**Comment by paddyroddy** [comment]
Okay, this is ready, always fiddly to add this kind of thing! Please review carefully as there are a fair few changes. They are generally trivial. I have self-reviewed and everything looks fine, I think. Have raised #248, #249 to follow up. This should probably me merged ASAP as other PRs will conflict with this potentially - and a fair bit of effort to fix.

File: pyproject.toml [hunk 4]
```diff
@@ -94,6 +94,60 @@ force-exclude = true
...
+    "ANN003", # TODO: missing-type-kwargs
+    "ANN201", # TODO: missing-return-type-undocumented-public-function
+    "COM812", # missing-trailing-comma (ruff-format recommended)
+    "D203", # one-blank-line-before-class
+    "D212", # blank-line-before-class
+    "ERA001", # TODO: commented-out-code
+    "ISC001", # single-line-implicit-string-concatenation (ruff-format recommended)
+    "NPY002", # TODO: numpy-legacy-random
+    "NPY201", # TODO: numpy2-deprecation
+    "RUF003", # ambiguous-unicode-character-comment
```
[the remark this answers is not in the document]
Comment (paddyroddy) [reply pyproject.toml:109]: #249

File: glass/core/__init__.py [hunk 5]
```diff
@@ -1,12 +1,12 @@
 # author: Nicolas Tessore <n.tessore@ucl.ac.uk>
 # license: MIT
 """
-Core functions (:mod:`glass.core`)
+Core functions (:mod:`glass.core`).
```
Comment (ntessore) [inline glass/core/__init__.py:4]: No full stops after headings

File: glass/core/array.py [hunk 6]
```diff
@@ -1,26 +1,27 @@
...
 def broadcast_leading_axes(*args):
-    """Broadcast all but the last N axes.
+    """
+    Broadcast all but the last N axes.
 
     Returns the shape of the broadcast dimensions, and all input arrays
     with leading axes matching that shape.
 
-    Example
+    Example:
```
Comment (ntessore) [inline glass/core/array.py:24]: This should probably "Examples" as in the [numpydoc format](https://numpydoc.readthedocs.io/en/latest/format.html#examples), but no `:` at the end.

File: glass/lensing.py [hunk 7]
```diff
@@ -439,12 +452,8 @@ def deflect(lon: ArrayLike, lat: ArrayLike, alpha: ArrayLike) -> NDArray:
     exponential map.
 
     """
-
     alpha = np.asanyarray(alpha)
-    if np.iscomplexobj(alpha):
-        alpha1, alpha2 = alpha.real, alpha.imag
-    else:
-        alpha1, alpha2 = alpha
+    alpha1, alpha2 = (alpha.real, alpha.imag) if np.iscomplexobj(alpha) else alpha
```
Comment (ntessore) [inline glass/lensing.py:456]: I would probably not understand the one-liner version of this on the first go.

File: glass/observations.py [hunk 8]
```diff
@@ -101,7 +107,7 @@ def gaussian_nz(
     ----------
     z : array_like
         Redshift values of the distribution.
-    mode : float or array_like
+    mean : float or array_like
```
Comment (ntessore) [inline glass/observations.py:110]: It's the mode, not the mean!

File: glass/shells.py [hunk 9]
```diff
@@ -120,17 +124,24 @@ def density_weight(z: ArrayLike, cosmo: "Cosmology") -> np.ndarray:
...
-RadialWindow.zeff.__doc__ = """Effective redshift of the window."""
+
+    za: Sequence[float]
+    wa: Sequence[float]
+    zeff: float
+
+
+RadialWindow.za.__doc__ = "Redshift array; the abscissae of the window function."
+RadialWindow.wa.__doc__ = "Weight array; the values (ordinates) of the window function."
+RadialWindow.zeff.__doc__ = "Effective redshift of the window."
```
Comment (ntessore) [inline glass/shells.py:135]: I think we can move these into the class and Sphinx will pick them up.

File: glass/user.py [hunk 10]
```diff
@@ -87,8 +93,10 @@ def write(self, data=None, /, **columns):
 @contextmanager
 def write_catalog(filename, *, ext=None):
     """
-    Write a catalogue into a FITS file, where *ext* is the optional
-    name of the extension.  To be used as a context manager::
+    Write a catalogue into a FITS file.
+
+    Where *ext* is the optional name of the extension.
```
Comment (ntessore) [inline glass/user.py:98]: Half sentence

File: tests/test_galaxies.py [hunk 11]
```diff
@@ -23,18 +26,19 @@ def test_redshifts():
...
 def test_redshifts_from_nz():
     import numpy as np
+
     from glass.galaxies import redshifts_from_nz
 
     # test sampling
 
     redshifts = redshifts_from_nz(10, [0, 1, 2, 3, 4], [1, 0, 0, 0, 0])
-    assert np.all((0 <= redshifts) & (redshifts <= 1))
+    assert np.all((redshifts >= 0) & (redshifts <= 1))
```
Comment (ntessore) [inline tests/test_galaxies.py:35]: No way, it's $a \le x \le b$.

**Review by ntessore — CHANGES_REQUESTED** [review]
Amazing piece of work! Had a quick look and other than the headings I think everything is a matter of taste.

File: glass/observations.py (hunk 8 above)
Comment (Saransh-cpp) [reply glass/observations.py:110]: Should the argument name in function be changed from `mean` to `mode`?

File: glass/shells.py [hunk 12]
```diff
@@ -1,8 +1,8 @@
 # author: Nicolas Tessore <n.tessore@ucl.ac.uk>
 # license: MIT
 """
-Shells
-======
+Shells.
+=======
```
Comment (Saransh-cpp) [inline glass/shells.py:5]:
```suggestion
Shells
======
```

File: glass/user.py [hunk 13]
```diff
@@ -1,8 +1,8 @@
 # author: Nicolas Tessore <n.tessore@ucl.ac.uk>
 # license: MIT
 """
-User utilities
-==============
+User utilities.
+===============
```
Comment (Saransh-cpp) [inline glass/user.py:5]:
```suggestion
User utilities
==============
```

File: glass/user.py (hunk 10 above)
Comment (Saransh-cpp) [inline glass/user.py:98]:
```suggestion
    *ext* is the optional name of the extension.
```

**Review by Saransh-cpp — COMMENTED** [review]
Taking over 🫡

**Review by Saransh-cpp — APPROVED** [review]
Amazing work, thanks! Looks good to me. I'll merge after the `mean`/`mode` confusion is cleared.

File: glass/observations.py (hunk 8 above)
Comment (ntessore) [reply glass/observations.py:110]: My mistake! I thought this was the other model (from which I guess the doctring was copied. `mean` is right!

**Review by ntessore — APPROVED** [review]
Excellent work @paddyroddy and @Saransh-cpp!

File: .pre-commit-config.yaml [hunk 14]
```diff
@@ -27,6 +27,7 @@ repos:
   - repo: https://github.com/astral-sh/ruff-pre-commit
     rev: v0.6.3
     hooks:
+      - id: ruff
```
[the remark this answers is not in the document]
Comment (paddyroddy) [reply .pre-commit-config.yaml:30]: Is `fix` not [covered here](https://github.com/glass-dev/glass/blob/285bda784074695cd3e55abc7101e4824165719d/pyproject.toml#L92)?

[the remark this answers is not in the document]
Comment (paddyroddy) [reply .pre-commit-config.yaml:30]: What does `--show-fixes` do? Should we move this to `pyproject.toml`?

**Comment by paddyroddy** [comment]
Thanks for taking over whilst I was away @Saransh-cpp!

File: .pre-commit-config.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (Saransh-cpp) [reply .pre-commit-config.yaml:30]: Oops, missed that. [`--show-fixes`](https://docs.astral.sh/ruff/settings/#show-fixes) is basically `-v` for the fixes ruff suggests.

[the remark this answers is not in the document]
Comment (paddyroddy) [reply .pre-commit-config.yaml:30]: I'll fix in #267

**Pull request merged**